Keywords

kevinboulain opened the pull request:
Hey there,

I saw that you implemented most of what was discussed in #25 so I did the remaining easy step and implemented the keywords :)

The actual commit implementing the keywords is the last one.
The two first are small changes that bothered me:
 - magit may not be (immediately) loaded/available, so try to load it (else feebleline may crash on Emacs' startup)
 - in some special buffers (like scratch) I lost my column and line numbers so I put them back

Feel free to cherry-pick only the ones you like.

By the way, the configuration I posted in #25 is now reduced to:
```elisp
(defun ether--feebleline-circe-tracking-buffers ()
  "Circe maintains a list of buffers with activity.
Join them together as they are already propertized in case of highlight."
  (when (boundp 'tracking-buffers)
    (string-join tracking-buffers " ")))

;; ...
  (setq feebleline-msg-functions
        '((feebleline-line-number :post "")
          (feebleline-column-number :pre ",")
          (feebleline-file-directory :face feebleline-dir-face :post "")
          (feebleline-file-or-buffer-name :face font-lock-keyword-face :post "")
          (feebleline-file-modified-star :face feebleline-warning-face :post "")
          (feebleline-git-object :face feebleline-git-face :pre ":")
          (ether--feebleline-circe-tracking-buffers :face nil :pre " ")))
;; ...
```

So that's a definitive improvement :)

I'll probably make another smaller pull request for the Circe function if you merge this one.

Regards.

PS: I'm still wondering if `feebleline-{line,column}-number` should respect `{line,column}-number-mode` that people may have set, so feebleline's behavior would match their mode-line. But I preferred not to include this decision here. But since, as a feebleline user, you may redefine the whole ` feebleline-msg-functions` that's probably unnecessary (and may introduce complexity, like how to handle an optional `pre` or `post` based on the previous or following function's output).

**EDIT**: the right alignment is still a bit suspicious, but since you already have comments about it I kept the same logic that you implemented (yes, it still works I tried with projectile enabled).

kevinboulain committed:
magit-get-current-branch may not be bound or return anything

It's probably better to try to import the required packages than to
check if it is bound because the user may not load magit itself.

Also fallback on rev-parse to display something in case the branch
isn't available (for example, detached HEAD).

kevinboulain committed:
use keywords instead of alist in feebleline-msg-functions

So it's a bit easier to customize.

kevinboulain committed:
always return line and column numbers

For example, buffer-file-name returns nothing for the scratch buffer,
which is probably not what we want.

tautologyclub commented:
Hey, cool! I'm looking at a few busy days, so I might not be able to check this out until later in the week, but thanks.

tautologyclub reviewed (commented):
In `feebleline--insert-func` you're referencing `tmp-string` which is let-bound inside `feebleline--insert`. Is this well-defined, if we're assuming `feebleline--insert-func` is only called by `feebleline--insert`? I mean it seems to work fine, I'd just like to know if functions "inheriting" let bindings is proper elisp.

That's the only comment I have on the keyword part, looks great other than that.

tautologyclub reviewed (commented):
I guess this makes more sense as a default, even though I prefer not showing it in non-file buffers.

tautologyclub reviewed (commented):
I don't think this is a good idea -- if you're not using magit, just remove the component from the function list.

Also, the way I did it on master branch was to wrap the whole defun with a soft require + fboundp. That way it's not checked every time. But I don't think there's any reason to get cute here - if people don't use magit they can customize the function list.

kevinboulain committed:
review: don't require magit

User should load magit if desired. Or simply remove the function from
feebleline-msg-functions.

kevinboulain commented:
Okay, so, in order:
 - yes, it's strange to use the let like so, but by default Emacs doesn't use lexical binding but dynamic binding, you can see an example with let in the [official documentation](https://www.gnu.org/software/emacs/manual/html_node/elisp/Dynamic-Binding.html#Dynamic-Binding). However, the main reason I did it like so was that it looked like (from the comments) that you may change it sooner than later because it doesn't handle every cases. By the way, did you think about maintaining two lists? The one to show on the left (so the functions not marked with `right-align`) and the one to show on the right (so the functions marked with `right-align`) that may easily solve your problem since you would simply have to concat the two with spaces in the middle. You could even extend it to an unique `:align` keyword with a value of `'left`, `'middle` and `'right` (so three lists internally).
 - for the lines numbers in every buffers, I'll let you choose, it's not like it's difficult to change anyway :)
 - I removed the `require`s on Magit, I admit it was probably a poor choice (but it allowed people to load it after feebleline). I replaced them with `fboundp` to avoid the requires but still avoid crashing during Emacs' startup if the functions aren't defined (that may a be a little bit better for a first contact with feebleline)

Don't hesitate to rewrite my code if you don't like it (or close the pull request)! I'm not here to push it on you :)

---

Another thought: unknown keywords could probably be forwarded to the formatter function so #26 could do something like (not tested):
```elisp
(cl-defun feebleline-time (&key (timefmt "%H:%M:%S"))
  ;; ...
)
(setq feebleline-msg-functions '((feebleline-time :fmt "%s" :timefmt "%H:%M"))) ; :timefmt being unknown it will be forwarded to feebleline-time
```
It could avoid the need to add too many custom variables.

endian-benjamin commented:
I'll get to reviewing this during the week - thanks a lot for the input and contributions so far!
